eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -0,0 +1,347 @@
+    )
+  }
+
+  /**
+   * This helper method translates relaying errors (returned by the downstream nodes) to a BOLT 4 standard error that we
+   * should return upstream.
+   */
+  def translateError(nodeParams: NodeParams, failures: Seq[PaymentFailure], upstream: Upstream.Trampoline, nextPayload: Onion.NodeRelayPayload): Option[FailureMessage] = {
+    val routeNotFound = failures.collectFirst { case f@LocalFailure(_, RouteNotFound) => f }.nonEmpty
+    val routingFeeHigh = upstream.amountIn - nextPayload.amountToForward >= nodeFee(nodeParams.feeBase, nodeParams.feeProportionalMillionth, nextPayload.amountToForward) * 5
+    failures match {
+      case Nil => None
+      case LocalFailure(_, BalanceTooLow) :: Nil if routingFeeHigh =>
+        // We have direct channels to the target node, but not enough outgoing liquidity to use those channels.
+        // The routing fee proposed by the sender was high enough to find alternative, indirect routes, but didn't yield
+        // any result so we tell them that we don't have enough outgoing liquidity at the moment.
+        Some(TemporaryNodeFailure)
+      case LocalFailure(_, BalanceTooLow) :: Nil => Some(TrampolineFeeInsufficient) // a higher fee/cltv may find alternative, indirect routes
+      case _ if routeNotFound => Some(TrampolineFeeInsufficient) // if we couldn't find routes, it's likely that the fee/cltv was insufficient
+      case _ =>
+        // Otherwise, we try to find a downstream error that we could decrypt.
+        val outgoingNodeFailure = failures.collectFirst { case RemoteFailure(_, e) if e.originNode == nextPayload.outgoingNodeId => e.failureMessage }
+        val otherNodeFailure = failures.collectFirst { case RemoteFailure(_, e) => e.failureMessage }
+        val failure = outgoingNodeFailure.getOrElse(otherNodeFailure.getOrElse(TemporaryNodeFailure))
+        Some(failure)
+    }
+  }
+
+}
+
+/**
+ * see https://doc.akka.io/docs/akka/current/typed/style-guide.html#passing-around-too-many-parameters
+ */
+class NodeRelay private(
+                         nodeParams: NodeParams,
+                         router: ActorRef,
+                         register: ActorRef,
+                         relayId: UUID,
+                         paymentHash: ByteVector32,
+                         context: ActorContext[NodeRelay.Command],
+                         adapters: typed.ActorRef[Any],
+                         fsmFactory: FsmFactory) {
+
+  import NodeRelay._
+
+
+  def apply(): Behavior[Command] =
+    Behaviors.receiveMessagePartial {
+      // We make sure we receive all payment parts before forwarding to the next trampoline node.
+      case WrappedNodeRelayPacket(IncomingPacket.NodeRelayPacket(add, outer, inner, next)) => outer.paymentSecret match {
+        case None =>
+          // TODO: @pm: maybe those checks should be done later in the flow (by the mpp FSM?)
```

Comment:
Maybe, it's worth exploring (in a separate PR, it's already quite hard to diff because of the file change!).
What's a bit tricky is that you want to reject HTLCs that don't have the right payment secret *without* impacting the state of the MPP FSM; you don't want to reject the other HTLCs (that have the right secret) because an attacker inserted a malicious HTLC that doesn't have it.